ads/google/a4a/google-data-reporter.js

```diff
@@ -0,0 +1,162 @@
+
+/**
+ * Check whether the element is in an experiment branch that is eligible for
+ * monitoring.
+ *
+ * @param {!AMP.BaseElement} ampElement
+ * @param {!string} namespace
+ * @returns {boolean}
+ */
+function isInReportableBranch(ampElement, namespace) {
+  // Handle the possibility of multiple eids on the element.
+  const eids = parseExperimentIds(
+      ampElement.element.getAttribute(EXPERIMENT_ATTRIBUTE));
+  const reportableA4AEids = {
+    [ADSENSE_A4A_EXTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+    [ADSENSE_A4A_INTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+    [DOUBLECLICK_A4A_EXTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+    [DOUBLECLICK_A4A_INTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+  };
+  const reportableControlEids = {
+    [ADSENSE_A4A_EXTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+    [ADSENSE_A4A_INTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+    [DOUBLECLICK_A4A_EXTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+    [DOUBLECLICK_A4A_INTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+  };
+  if (namespace == 'a4a' &&
+      (eids.some(x => { return x in reportableA4AEids; }) ||
+      isInManualExperiment(ampElement.element))) {
+    return true;
+  } else if (namespace == 'amp' &&
+      eids.some(x => { return x in reportableControlEids; })) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+/**
+ * @param {!AMP.BaseElement} ampElement The element on whose lifecycle this
+ *    reporter will be reporting.
+ * @param {string} namespace
+ * @param {number|string} slotId A unique numeric identifier in the page for
+ *    the given element's slot.
+ * @return {!GoogleAdLifecycleReporter|!BaseLifecycleReporter}
+ */
+export function getLifecycleReporter(ampElement, namespace, slotId) {
+  // Carve-outs: We only want to enable profiling pingbacks when:
+  //   - The ad is from one of the Google networks (AdSense or Doubleclick).
+  //   - The ad slot is in the A4A-vs-3p amp-ad control branch (either via
+  //     internal, client-side selection or via external, Google Search
+  //     selection).
+  //   - We haven't turned off profiling via the rate controls in
+  //     build-system/global-config/{canary,prod}-config.json
+  // If any of those fail, we use the `BaseLifecycleReporter`, which is a
+  // a no-op (sends no pings).
+  const type = ampElement.element.getAttribute('type');
+  const win = ampElement.win;
+  // In local dev mode, neither the canary nor prod config files is available,
+  // so manually set the profiling rate, for testing/dev.
+  if (getMode().localDev &&
+      (!win.AMP_CONFIG || !win.AMP_CONFIG['a4aProfilingRate'])) {
+    win.AMP_CONFIG = win.AMP_CONFIG || {};
+    win.AMP_CONFIG['a4aProfilingRate'] = 1.0;
+  }
+  randomlySelectUnsetPageExperiments(win, win.AMP_CONFIG['a4aProfilingRate']);
+  if ((type == 'doubleclick' || type == 'adsense') &&
+      isInReportableBranch(ampElement, namespace) &&
+      isExperimentOn(win, 'a4aProfilingRate')) {
+    const correlator = getCorrelator(win);
```

Comment:
Why do we need to pass correlator?  Given getCorrelator is imported and only requires win (which is also passed to the constructor) can't this be populated from within the GoogleAdLifecycleReporter constructor?